src/libcrun/cgroup.c

```diff
@@ -736,7 +738,73 @@ systemd_job_removed (sd_bus_message *m, void *userdata, sd_bus_error *error arg_
     return -1;
 
   if (strcmp (p->path, path) == 0)
-    *p->terminated = 1;
+    {
+      p->terminated = 1;
+      p->unit = xstrdup (unit);
+      p->result = xstrdup (result);
+    }
+  return 0;
+}
+
+static int
+systemd_check_job_status_setup (sd_bus *bus,
+                                void **data,
+                                libcrun_error_t *err)
+{
+  int ret;
+
+  *data = xmalloc0 (sizeof (struct systemd_job_removed_s));
+  ret = sd_bus_match_signal_async (bus,
+                                   NULL,
+                                   "org.freedesktop.systemd1",
+                                   "/org/freedesktop/systemd1",
+                                   "org.freedesktop.systemd1.Manager",
+                                   "JobRemoved",
+                                   systemd_job_removed,
+                                   NULL,
+                                   *data);
+  if (UNLIKELY (ret < 0))
+    return crun_make_error (err, -ret, "sd-bus match signal");
+
+  return 0;
+}
+
+static int
+systemd_check_job_status (sd_bus *bus,
+                          void *data,
+                          const char *path,
+                          const char *op,
+                          libcrun_error_t *err)
+{
+  int sd_err;
+  cleanup_free struct systemd_job_removed_s *d = data;
+
+  d->path = path;
+  while (!d->terminated)
+    {
+      sd_err = sd_bus_process (bus, NULL);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus process");
+
+      if (sd_err == 0)
+        {
+          sd_err = sd_bus_wait (bus, (uint64_t) -1);
+          if (UNLIKELY (sd_err < 0))
+            return crun_make_error (err, -sd_err, "sd-bus wait");
+
+          continue;
+        }
+    }
+
+  if (d->unit && d->result)
+    {
+      if (strcmp (d->result, "done") != 0)
+          return crun_make_error (err, 0, "error %s systemd unit %s: %s", op, d->unit, d->result);
```

Comment:
could we make sure we don't leak `d->unit` and `d->result` here?

We could just store the return code instead of returning here.

If you feel like playing with autocleanup (but absolutely not necessary), another alternative could be something like: https://github.com/containers/crun/blob/master/src/libcrun/linux.c#L1556 and define a function to do the cleanup for data instead of `cleanup_free`, and the function could free both strings and the struct itself.

Reply:
Waaaay too much golang for me :sob: